Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean

```diff
@@ -101,21 +101,235 @@ lemma isHamiltonianCycle_iff_isCycle_and_support_count_tail_eq_one :
 lemma IsHamiltonianCycle.mem_support (hp : p.IsHamiltonianCycle) (b : α) :
     b ∈ p.support := List.mem_of_mem_tail <| support_tail p _ ▸ hp.isHamiltonian_tail.mem_support _
 
+namespace IsHamiltonianCycle
+
+open scoped List
+
 /-- The length of a hamiltonian cycle is the number of vertices. -/
-lemma IsHamiltonianCycle.length_eq [Fintype α] (hp : p.IsHamiltonianCycle) :
+lemma length_eq [Fintype α] (hp : p.IsHamiltonianCycle) :
     p.length = Fintype.card α := by
   rw [← length_tail_add_one hp.not_nil, hp.isHamiltonian_tail.length_eq, Nat.sub_add_cancel]
   rw [Nat.succ_le, Fintype.card_pos_iff]
   exact ⟨a⟩
 
-lemma IsHamiltonianCycle.count_support_self (hp : p.IsHamiltonianCycle) :
+lemma count_support_self (hp : p.IsHamiltonianCycle) :
     p.support.count a = 2 := by
   rw [support_eq_cons, List.count_cons_self, ← support_tail, hp.isHamiltonian_tail]
 
-lemma IsHamiltonianCycle.support_count_of_ne (hp : p.IsHamiltonianCycle) (h : a ≠ b) :
+lemma support_count_of_ne (hp : p.IsHamiltonianCycle) (h : a ≠ b) :
     p.support.count b = 1 := by
   rw [← cons_support_tail p, List.count_cons_of_ne h.symm, hp.isHamiltonian_tail]
 
+protected theorem transfer (hp : p.IsHamiltonianCycle) {H} (h) :
+    (p.transfer H h).IsHamiltonianCycle := by
+  rw [isHamiltonianCycle_iff_isCycle_and_support_count_tail_eq_one] at *
+  refine And.intro (hp.1.transfer _) (fun a => ?_)
+  simp only [support_transfer]
+  exact hp.2 a
+
+variable (b)
+
+protected theorem rotate (hp : p.IsHamiltonianCycle) :
+    IsHamiltonianCycle (p.rotate (hp.mem_support b)) := by
+  rw [Walk.isHamiltonianCycle_iff_isCycle_and_support_count_tail_eq_one] at *
+  refine And.intro ?_ (fun v => ?_)
+  · apply hp.1.rotate
+  · rw [← hp.2 v]
+    exact List.perm_iff_count.mp (List.IsRotated.perm (by simp)) v
+
+lemma mem_tail_support (hp : p.IsHamiltonianCycle) : b ∈ p.support.tail := by
+  rw [← List.count_pos_iff_mem]
+  have := hp.2 b
+  simp at this
+  omega
+
+lemma mem_dropLast_support (hp : p.IsHamiltonianCycle) : b ∈ p.support.dropLast := by
+  rw [List.IsRotated.mem_iff (IsRotated_dropLast_tail p)]
+  apply hp.mem_tail_support
+
+/-- The dart in the Hamiltonian cycle that starts at `b` -/
+noncomputable def dart_with_fst (hp : p.IsHamiltonianCycle) : G.Dart :=
```

Comment:
I would advise you make this definition (along with the following ones) a separate PR

Reply:
But this may make the dependency of PR more complex...